### ISSUE TITLE
Saved State on Disk is very large

### ISSUE THREAD

**Issue by tc-wolf** [abetlen/llama-cpp-python#1303]
**Is your feature request related to a problem? Please describe.**
When using `Llama.save_state`, the size on disk is very large, partially because `Llama.scores` is always of size `(n_ctx, n_vocab)`, even when the number of tokens actually used is much less than this.

**Describe the solution you'd like**
* Only store the number of tokens used by the model (`self.n_tokens`).
* If possible, re-use logits from the `llama_context` struct instead of serializing `scores`.
  * These didn't match when I loaded both and compared, though didn't have logit processor callbacks.

**Describe alternatives you've considered**
* Somehow compressing `LlamaState`
* Re-using logits from C++ state saving functions and eliminating `scores` on Python model
  * Looks like this is needed for sampling logic

Partially fixed by #1296

**Comment by tc-wolf** [comment]
Closing bc #1296 merged
